external-plugins/rehearse/plugin/handler/handler.go

```diff
@@ -0,0 +1,463 @@
+package handler
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	v1 "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
+
+	"github.com/sirupsen/logrus"
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	gitv2 "k8s.io/test-infra/prow/git/v2"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+var log *logrus.Logger
+
+func init() {
+	log = logrus.New()
+	log.SetOutput(os.Stdout)
+}
+
+// GitHubEvent represents a valid GitHub event in the events channel
+type GitHubEvent struct {
+	Type    string
+	GUID    string
+	Payload []byte
+}
+
+type GitHubEventsHandler struct {
+	eventsChan       <-chan *GitHubEvent
+	logger           *logrus.Logger
+	prowClient       v1.ProwJobInterface
+	ghClient         githubClientInterface
+	gitClientFactory gitv2.ClientFactory
+	prowConfigPath   string
+	jobsConfigBase   string
+	alwaysRun        bool
+}
+
+// NewGitHubEventsHandler returns a new github events handler
+func NewGitHubEventsHandler(
+	eventsChan <-chan *GitHubEvent,
+	logger *logrus.Logger,
+	prowClient v1.ProwJobInterface,
+	ghClient githubClientInterface,
+	prowConfigPath string,
+	jobsConfigBase string,
+	alwaysRun bool,
+	gitClientFactory gitv2.ClientFactory) *GitHubEventsHandler {
+
+	return &GitHubEventsHandler{
+		eventsChan:       eventsChan,
+		logger:           logger,
+		prowClient:       prowClient,
+		ghClient:         ghClient,
+		prowConfigPath:   prowConfigPath,
+		jobsConfigBase:   jobsConfigBase,
+		alwaysRun:        alwaysRun,
+		gitClientFactory: gitClientFactory,
+	}
+}
+
+type githubClientInterface interface {
+	IsMember(string, string) (bool, error)
+	GetPullRequest(string, string, int) (*github.PullRequest, error)
+}
+
+func (h *GitHubEventsHandler) Handle(incomingEvent *GitHubEvent) {
+
+	log.Infoln("GitHub events handler started")
+	eventLog := log.WithField("event-guid", incomingEvent.GUID)
+	switch incomingEvent.Type {
+	case "pull_request":
+		eventLog.Infoln("Handling pull request event")
+		var event github.PullRequestEvent
+		if err := json.Unmarshal(incomingEvent.Payload, &event); err != nil {
+			eventLog.WithError(err).Error("Could not unmarshal event.")
+			return
+		}
+		h.handlePullRequestUpdateEvent(eventLog, &event)
```

Comment:
What about returning errors here and composing them with `github.com/pkg/errors` using `errors.Wrap` and friends ? we will do the logrus.WithError just here.

Reply:
Who should handle the error that is returned?

Reply:
I suppose the border is here https://github.com/kubevirt/project-infra/pull/554/files#r464933931